sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/interceptor/SasTokenCredentialInterceptor.java

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.storage.blob.interceptor;
+
+import com.azure.android.storage.blob.credentials.SasTokenCredential;
+
+import java.io.IOException;
+
+import okhttp3.HttpUrl;
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+
+/**
+ *  Interceptor that append SAS token to the request Uri.
+ */
+public class SasTokenCredentialInterceptor implements Interceptor {
+    private final SasTokenCredential credential;
+
+    public SasTokenCredentialInterceptor(SasTokenCredential credential) {
+        this.credential = credential;
+    }
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        HttpUrl requestURL = chain.request().url();
+        String delimiter = (requestURL.query() != null && requestURL.query() != "") ? "&" : "?";
+        requestURL.toString();
+        String newURL = requestURL.toString() + delimiter + this.credential.getSasToken();
```

Comment:
This could have a bug if the query string is not the last part of the request URL. For example, a URL can have a `#fragment` component.

Reply:
We can do something like this to be more defensive: 

```java
String sasToken = this.credential.getSasToken();
if (sasToken.startsWith("?")) {
    sasToken = sasToken.substring(1);
}

HttpUrl sasDummyUrl = new HttpUrl.Builder()
        .scheme("http")
        .host("localhost")
        .query(sas)
        .build();

HttpUrl.Builder newUrlBuilder = chain.request().url().newBuilder();
for (String qName : sasDummyUrl.queryParameterNames()) {
    newUrlBuilder.addEncodedQueryParameter(qName, httpUrl.queryParameter(qName));
}

Request newRequest = request
        .newBuilder()
        .url(newUrlBuilder.build())
        .build();

return chain.proceed(newRequest);
```

The fragment part is not something supported or has no meaning in storage blob/container space, so the chances of hitting a bug in the current logic is very low. The defensive logic has extra parsing and looping involved per request, so a minor concern is does it add unnecessary overhead.

Thoughts?

Reply:
We can do something like this to be more defensive: 

```java
String sasToken = this.credential.getSasToken();
if (sasToken.startsWith("?")) {
    sasToken = sasToken.substring(1);
}

HttpUrl sasDummyUrl = new HttpUrl.Builder()
        .scheme("http")
        .host("localhost")
        .query(sas)
        .build();

HttpUrl.Builder newUrlBuilder = chain.request().url().newBuilder();
for (String qName : sasDummyUrl.queryParameterNames()) {
    newUrlBuilder.addEncodedQueryParameter(qName, httpUrl.queryParameter(qName));
}

Request newRequest = request
        .newBuilder()
        .url(newUrlBuilder.build())
        .build();

return chain.proceed(newRequest);
```

The fragment part is not something supported or has no meaning in storage blob/container space, so the chances of hitting a bug in the current logic is very low. The defensive logic has extra parsing and looping involved per request, so a minor concern is does it add unnecessary overhead.

Thoughts?

Reply:
How would you feel about:

```java
public Response intercept(Chain chain) throws IOException {
    HttpUrl requestURL = chain.request().url();
    String query = requestURL.query();
    if (!CoreUtils.isNullOrEmpty(query)) {
        query += "&";
    }
    query += this.credential.getSasToken();
    HttpUrl newURL = requestURL
        .newBuilder()
        .query(query)
        .build();
    Request newRequest = chain.request()
            .newBuilder()
            .url(newURL)
            .build();
    return chain.proceed(newRequest);
}
```

Reply:
thank you, this looks very good. I think we just needed to add "?" trimming in SaS token. Let me incorporate it and test (specifically encoding part), if all good I will update the PR with it and the other one to import CallBack.

Reply:
The `requestURL.query()` will give "decoded query string". 
The `this.credential.getSasToken()` will give "encoded query string".

So doing `requestURL.query(decoded query string + encoded query string)` will be a problem. The `query(..)` method expect a decoded value and does encoding internally.

It looks like either we need 

```java
HttpUrl sasDummyUrl = new HttpUrl.Builder()
        .scheme("http")
        .host("localhost")
        .query(sas)
        .build();
``` 

or we need to manually split the sas by `&` then each part by `=`

Finally use `addEncodedQueryParameter`.

Reply:
It looks like both `HttpUrl` and `HttpUrl.Builder` have equivalent `encodedQuery()` methods, that should fix the problem right?

```java
public Response intercept(Chain chain) throws IOException {
    HttpUrl requestURL = chain.request().url();
    String encodedQuery = requestURL.encodedQuery();
    String sasToken = this.credential.getSasToken();

    if (!CoreUtils.isNullOrEmpty(encodedQuery)) {
        encodedQuery += "&";
    }
    encodedQuery += sasToken.startsWith("?")
        ? sasToken.substring(1)
        : sasToken;
    HttpUrl newURL = requestURL
        .newBuilder()
        .encodedQuery(encodedQuery)
        .build();
    Request newRequest = chain.request()
            .newBuilder()
            .url(newURL)
            .build();
    return chain.proceed(newRequest);
}
```

Reply:
Good catch, I missed `requestURL.encodedQuery()` API. thanks Brandon!.

Reply:
Honestly, I was 75% expecting those APis not to be exposed, forcing us to have to resort to something more verbose. I saw that there is a lot of useful, general-purpose URL-manipulation functionality present inside the `HttpUrl` class but sadly we can't use the vast majority of it because all the methods are package-private and the class itself is final.

Reply:
right, OkHttp has many good things hidden, but as internal. I wish they publish it as a separate light weight packages (buffer, url-utils etc..).